syft/source/directory_source.go

```diff
@@ -0,0 +1,215 @@
+package source
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/bmatcuk/doublestar/v4"
+	"github.com/opencontainers/go-digest"
+
+	"github.com/anchore/syft/internal/log"
+	"github.com/anchore/syft/syft/artifact"
+	"github.com/anchore/syft/syft/file"
+	"github.com/anchore/syft/syft/internal/fileresolver"
+)
+
+var _ Source = (*DirectorySource)(nil)
+
+type DirectoryConfig struct {
+	Path    string
+	Base    string
+	Exclude ExcludeConfig
+	Alias   *Alias
+}
+
+type DirectorySourceMetadata struct {
+	Path string `json:"path" yaml:"path"`
+	Base string `json:"-" yaml:"-"` // though this is important, for display purposes it leaks too much information (abs paths)
+}
+
+type DirectorySource struct {
+	id       artifact.ID
+	config   DirectoryConfig
+	resolver *fileresolver.Directory
+	mutex    *sync.Mutex
+}
+
+func NewFromDirectoryPath(path string) (*DirectorySource, error) {
+	cfg := DirectoryConfig{
+		Path: path,
+	}
+	return NewFromDirectory(cfg)
+}
+
+func NewFromDirectory(cfg DirectoryConfig) (*DirectorySource, error) {
+	fi, err := os.Stat(cfg.Path)
+	if err != nil {
+		return nil, fmt.Errorf("unable to stat path=%q: %w", cfg.Path, err)
+	}
+
+	if !fi.IsDir() {
+		return nil, fmt.Errorf("given path is not a directory: %q", cfg.Path)
+	}
+
+	return &DirectorySource{
+		id:     deriveIDFromDirectory(cfg),
+		config: cfg,
+		mutex:  &sync.Mutex{},
+	}, nil
+}
+
+// deriveIDFromDirectory generates an artifact ID from the given directory config. If an alias is provided, then
+// the artifact ID is derived exclusively from the alias name and version. Otherwise, the artifact ID is derived
+// from the path provided with an attempt to prune a prefix if a base is given. Since the contents of the directory
+// are not considered, there is no semantic meaning to the artifact ID -- this is why the alias is preferred without
+// consideration for the path.
+func deriveIDFromDirectory(cfg DirectoryConfig) artifact.ID {
+	var info string
+	if cfg.Alias != nil {
+		// don't use any of the path information -- instead use the alias name and version as the artifact ID.
+		// why? this allows the user to set a dependable stable value for the artifact ID in case the
+		// scanning root changes (e.g. a user scans a directory, then moves it to a new location and scans again).
+		info = fmt.Sprintf("%s@%s", cfg.Alias.Name, cfg.Alias.Version)
+	} else {
+		log.Warn("no explicit name and version provided for directory source, deriving artifact ID from the given path (which is not ideal)")
```

Comment:
I wanted to call out this change explicitly -- I still haven't been able to reconcile what a "good" ID is for a directory (that isn't too expensive to figure). But I've added this warning to at least let folks know that if you're doing a directory scan an are not setting `--source-name` or `--source-version` explicitly then this is most likely wrong. I didn't call out the specific options since this is the lib path and not in the cmd pkg.

Reply:
should this be an Info or Debug message? This will show for _every_ directory scan otherwise, I think

Reply:
I think that's the problem with the directory source ID -- it's really not adequate for any use case where you're trying to cross correlate SBOMs historically since it's brittle and may overlap with entirely unrelated scans. It feels like it should be a warning for that reason.

Reply:
I did find that this wasn't showing up though, and I discovered why: we can't assume that a pointer to alias means that the alias is optional, we still need to check individual fields. I've updated the API to not take alias pointers in config objects (any spot on the exported API where they were function args it is still a pointer though). This allows for the zero value to be used in configs (or for the caller to fully specify like with the cmd package) and we need to check the name and version explicitly, which was already being done in some circumstances.